OpenRA.Game/Activities/Activity.cs

```diff
@@ -9,50 +9,216 @@
  */
 #endregion
 
+using System;
 using System.Collections.Generic;
 using System.Linq;
 using OpenRA.Traits;
 
 namespace OpenRA.Activities
 {
+	public enum ActivityState { Queued, Active, Done, Canceled }
+
 	public abstract class Activity
 	{
-		public Activity NextActivity { get; set; }
+		public ActivityState State { get; private set; }
+
+		public Activity Root
+		{
+			get
+			{
+				if (Parent == null)
+					return this;
+
+				var p = Parent;
+				while (p.Parent != null)
+					p = p.Parent;
+
+				return p;
+			}
+		}
+
+		private Activity parent;
+		public Activity Parent
+		{
+			get
+			{
+				return parent;
+			}
+
+			protected set
+			{
+				parent = value;
+
+				var next = nextActivity;
+				if (next != null)
+					next.Parent = parent;
+			}
+		}
+
+		private Activity childActivity;
+		protected Activity ChildActivity
+		{
+			get
+			{
+				return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null;
+			}
+
+			set
+			{
+				if (value == this || value == Parent || value == nextActivity)
```

Comment:
Yup, they wouldn't be caught, possibly creating a loop. Checking more of the tree is quite a difficult and probably expensive operation. You'd not only have to check the parent's parents, but the parent's next queue as well (and the parent's parent's next queue, and their parents and so on). I went with just checking for those activities that would be likely to be returned by the activities own `Tick()` method instead.